### PR TITLE
redo custom feature

### PR DESCRIPTION
**Pull request by CoolLoong**

### CODE REVIEW

File: src/test/java/cn/nukkit/item/TestCustomItem.java [hunk 1]
```diff
@@ -1,9 +1,7 @@
-package cn.nukkit.item;
 
-import cn.nukkit.item.customitem.ItemCustom;
-import org.graalvm.polyglot.Context;
-import org.graalvm.polyglot.Value;
+package cn.nukkit.item;
 
+/*
```
Comment (Superice666) [inline src/test/java/cn/nukkit/item/TestCustomItem.java:4]: 这里不要注释掉它，这个类存在能被加载成功就是它的测试用途。

**Review by Superice666 — CHANGES_REQUESTED** [review]
改动过大，建议将完整迁移文档一并提交后再合并

File: src/test/java/cn/nukkit/item/TestCustomItem.java (hunk 1 above)
Comment (CoolLoong) [reply src/test/java/cn/nukkit/item/TestCustomItem.java:4]: 自定义物品部分更改，所以这里需要修改。
后面我加上完整的javadoc和完整迁移文档

Comment (Superice666) [reply src/test/java/cn/nukkit/item/TestCustomItem.java:4]: 这个类存在是为了检验JS API背后的字节码分析器能正常工作

**Comment by CoolLoong** [comment]
### Custom Feature Api migration documentation
All custom feature need to implement the `public CustomItemDefinition getDefinition()` method,which corresponds to `CustomItemDefinition` `CustomBlockDefinition` `CustomEntityDefinition` respectively.These three are all simple builder tool classes.
#### Api Remove

- `Entity#registerCustomEntity(CustomEntityDefinition customEntityDefinition, Class<? extends Entity> entity)`
   Replace with
  `Entity#registerCustomEntity(CustomEntityProvider customEntityProvider)`

- `CustomBlock#getTexture()`
   Replace with
  `CustomBlockDefinition.builder(this, Texture Name)` in `CustomBlock#getDefinition()`

- `CustomBlock#getMaterials()`
  Replace with
  `CustomBlockDefinition.builder(this, Materials materials,BlockCreativeCategory blockCreativeCategory)` in 
  `CustomBlock#getDefinition()`

- `CustomBlock#getGeometry()`
  Replace with
  `CustomBlockDefinition.Builder#geometry(String geometry)` in `CustomBlock#getDefinition()`

- `CustomBlock#componentNBTProcessor(CompoundTag componentNBT)`
  Replace with
  `CustomBlockDefinition.Builder#customBuild(Consumer<CompoundTag> nbt)` in `CustomBlock#getDefinition()`

- `CustomBlock#getPermutations()`
  Replace with
  `CustomBlockDefinition.Builder#permutations(Permutations permutations)` in `CustomBlock#getDefinition()`
#### Api Changed

- `CustomBlock#getNamespace() -> CustomBlock#getNamespaceId()`
#### Example of a simple pickaxe
```java
public class MyPickaxe extends ItemCustomTool {
    public MyPickaxe() {
        super("blocklynukkit:my_pickaxe", "测试镐", "amethyst_pickaxe");

    }

    @Override
    public CustomItemDefinition getDefinition() {
        return CustomItemDefinition
                .toolBuilder(this, ItemCreativeCategory.EQUIPMENT)
                .speed(10)//Set the digging speed of the pickaxe
                .addExtraBlockTag(BlockState.of("minecraft:nether_brick").getBlock(), 10)//Add additional matching block, if the block do not match then it is normal digging speed
                .allowOffHand(true)//Can be held in the offHand
                .handEquipped(true)//If true is the way the tool is displayed, false is the item
                .foil(true)//Items with enchanted light
                .build();
    }

    @Override
    public int getMaxDurability() {
        return ItemTool.DURABILITY_STONE;
    }

    @Override
    public boolean isPickaxe() {
        return true;
    }

    @Override
    public int getAttackDamage() {
        return 3;
    }

    @Override
    public int getEnchantAbility() {
        return 50;
    }

    @Override
    public int getTier() {
        return ItemCustomTool.TIER_NETHERITE;
    }
}
```

#### Example of a custom slab
```java
public class MySlab extends BlockTransparentMeta implements CustomBlock {
    public final BooleanBlockProperty BRIDGE_TOP_SLOT_BIT = new BooleanBlockProperty("bridge:top_slot_bit", false);
    public final BooleanBlockProperty BRIDGE_IS_FULL_BIT = new BooleanBlockProperty("bridge:is_full_bit", false);

    @Override
    public String getNamespaceId() {
        return "blocklynukkit:blue_mahoe_slab";
    }

    @Override
    public String getName() {
        return CustomBlock.super.getName();
    }

    @Override
    public int getId() {
        return CustomBlock.super.getId();
    }

    @NotNull
    @Override
    public BlockProperties getProperties() {
        return new BlockProperties(
                BRIDGE_TOP_SLOT_BIT, BRIDGE_IS_FULL_BIT
        );
    }

    @Override
    public CustomBlockDefinition getDefinition() {
        return CustomBlockDefinition
                .builder(this, "blue_mahoe_planks")//texture name
                .geometry("geometry.custom_slab")//geometry name
                .permutations(new Permutations(//permutation data
                        Permutation.builder()
                                .collision_box_enabled(true)
                                .collision_box_origin(new Vector3f(-8, 0, -8))
                                .collision_box_size(new Vector3f(16, 8, 16))
                                .selection_box_enabled(true)
                                .selection_box_origin(new Vector3f(-8, 0, -8))
                                .selection_box_size(new Vector3f(16, 8, 16))
                                .condition("query.block_property('bridge:top_slot_bit') == false && query.block_property('bridge:is_full_bit') == false"),
                        Permutation.builder()
                                .collision_box_enabled(true)
                                .collision_box_origin(new Vector3f(-8, 8, -8))
                                .collision_box_size(new Vector3f(16, 16, 16))
                                .selection_box_enabled(true)
                                .selection_box_origin(new Vector3f(-8, 8, -8))
                                .selection_box_size(new Vector3f(16, 16, 16))
                                .condition("query.block_property('bridge:top_slot_bit') == true && query.block_property('bridge:is_full_bit') == false"),
                        Permutation.builder()
                                .collision_box_enabled(true)
                                .collision_box_origin(new Vector3f(-8, 0, -8))
                                .collision_box_size(new Vector3f(16, 16, 16))
                                .selection_box_enabled(true)
                                .selection_box_origin(new Vector3f(-8, 0, -8))
                                .selection_box_size(new Vector3f(16, 16, 16))
                                .condition("query.block_property('bridge:is_full_bit') == true")
                ))
                .customBuild((nbt) -> {//custom processing result compound
                    nbt.getCompound("components").putCompound("minecraft:part_visibility", new CompoundTag()
                            .putCompound("boneConditions", new CompoundTag()
                                    .putCompound("lower", new CompoundTag()
                                            .putString("bone_condition", "!query.block_property('bridge:top_slot_bit') || query.block_property('bridge:is_full_bit')")
                                            .putString("bone_name", "lower")
                                            .putInt("molang_version", 6))
                                    .putCompound("upper", new CompoundTag()
                                            .putString("bone_condition", "query.block_property('bridge:top_slot_bit') || query.block_property('bridge:is_full_bit')")
                                            .putString("bone_name", "upper")
                                            .putInt("molang_version", 6))));
                });
    }
    //Block digging takes time
    @Override
    public double calculateBreakTime(@NotNull Item item, @Nullable Player player) {
        return 3;
    }
    //Block friction factor
    @Override
    public double getFrictionFactor() {
        return 0.1;
    }
    //Block explosion resistance
    @Override
    public double getResistance() {
        return 5;
    }
    //Block emit light level
    @Override
    public int getLightLevel() {
        return 15;
    }
    //Block absorb light level
    @Override
    public int getLightFilter() {
        return 15;
    }
    //burn ability
    @Override
    public int getBurnAbility() {
        return 0;
    }
    //burn chance
    @Override
    public int getBurnChance() {
        return 0;
    }
    //Block item max stack size
    @Override
    public int getItemMaxStackSize() {
        return 64;
    }

    @Override
    public boolean place(@NotNull Item item, @NotNull Block block, @NotNull Block target, @NotNull BlockFace face, double fx, double fy, double fz, @Nullable Player player) {
        //....Server-side processing behavior
    }
}
```

#### For change details, please see
https://github.com/PowerNukkitX/ExamplePlugin-Experiment-Mode/pull/1/files

**Review by FrioGitHub — CHANGES_REQUESTED** [review]
Very useful.

**Review by lt-name — APPROVED** [review]

File: src/main/java/cn/nukkit/item/customitem/ItemCustom.java [hunk 2]
```diff
@@ -2,128 +2,42 @@
...
-        }
-        this.runtimeId = INTERNAL_ALLOCATION_ID_MAP.get(this.getNamespaceId());
     }
 
     public ItemCustom(@Nonnull String id, @Nullable String name, @Nonnull String textureName) {
         this(id, name);
         this.textureName = textureName;
     }
 
-    public boolean allowOffHand() {
```
Comment (lt-name) [inline src/main/java/cn/nukkit/item/customitem/ItemCustom.java:55]: 这个好像是nk内部判断也要用的吧

**Review by lt-name — APPROVED** [review]

File: src/main/java/cn/nukkit/block/Block.java [hunk 3]
```diff
@@ -1218,13 +1219,31 @@ public static void registerBlockImplementation(int blockId, @Nonnull Class<? ext
      * @param blockClassList 传入自定义方块class List
      */
     @PowerNukkitXOnly
-    @Deprecated(since = "1.19.20-r5")
     public static void registerCustomBlock(@Nonnull List<Class<? extends CustomBlock>> blockClassList) throws NoSuchMethodException, InvocationTargetException, InstantiationException, IllegalAccessException {
```
Comment (lt-name) [inline src/main/java/cn/nukkit/block/Block.java:1222]: 我觉得可以从两个registerCustomBlock方法中提取出来通用部分

File: src/test/java/cn/nukkit/item/TestCustomItem.java (hunk 1 above)
Comment (CoolLoong) [reply src/test/java/cn/nukkit/item/TestCustomItem.java]: fixed

**Pull request merged**